### ISSUE TITLE
cisco switch connection failed.

### ISSUE THREAD

**Issue by zhanglinqiang** [mwiede/jsch#94]
Hello, I'm back again. I have two cisco switches that failed to connect. Can you help me see what's the problem?

jsch Version (with bug fixed #90 ) ：
```
<dependency>
            <groupId>com.github.mwiede</groupId>
            <artifactId>jsch</artifactId>
            <version>0.1.69</version>
        </dependency>
```
java version:
 ```
java version "1.8.0_221"
Java(TM) SE Runtime Environment (build 1.8.0_221-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.221-b11, mixed mode)
```
stack trace:
```
com.jcraft.jsch.JSchException: Session.connect: java.io.IOException: End of IO Stream Read
	at com.jcraft.jsch.Session.connect(Session.java:570)
	at com.topsec.tap.collector.agent.connection.JschConnection.initializeOpen(JschConnection.java:217)
	at com.topsec.tap.collector.agent.connection.JschConnection.open(JschConnection.java:140)
	at com.topsec.tap.collector.agent.connection.Connection$1.execute(Connection.java:68)
	at com.topsec.tap.collector.agent.connection.Connection$1.execute(Connection.java:64)
	at com.topsec.tap.collector.agent.util.TimeoutExecution.call(TimeoutExecution.java:61)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: End of IO Stream Read
	at com.jcraft.jsch.IO.getByte(IO.java:85)
	at com.jcraft.jsch.Session.read(Session.java:1178)
	at com.jcraft.jsch.Session.connect(Session.java:314)
	... 11 more
```
ssh info (one):
```
[root@localhost system]# ssh -vvv 10.28.11.41
OpenSSH_8.6p1, OpenSSL 1.1.1k  25 Mar 2021
debug1: Reading configuration data /etc/ssh/ssh_config
debug2: resolve_canonicalize: hostname 10.28.11.41 is address
debug3: expanded UserKnownHostsFile '~/.ssh/known_hosts' -> '/root/.ssh/known_hosts'
debug3: expanded UserKnownHostsFile '~/.ssh/known_hosts2' -> '/root/.ssh/known_hosts2'
debug1: Authenticator provider $SSH_SK_PROVIDER did not resolve; disabling
debug3: ssh_connect_direct: entering
debug1: Connecting to 10.28.11.41 [10.28.11.41] port 22.
debug3: set_sock_tos: set socket 3 IP_TOS 0x48
debug1: Connection established.
debug1: identity file /root/.ssh/id_rsa type -1
debug1: identity file /root/.ssh/id_rsa-cert type -1
debug1: identity file /root/.ssh/id_dsa type -1
debug1: identity file /root/.ssh/id_dsa-cert type -1
debug1: identity file /root/.ssh/id_ecdsa type -1
debug1: identity file /root/.ssh/id_ecdsa-cert type -1
debug1: identity file /root/.ssh/id_ecdsa_sk type -1
debug1: identity file /root/.ssh/id_ecdsa_sk-cert type -1
debug1: identity file /root/.ssh/id_ed25519 type -1
debug1: identity file /root/.ssh/id_ed25519-cert type -1
debug1: identity file /root/.ssh/id_ed25519_sk type -1
debug1: identity file /root/.ssh/id_ed25519_sk-cert type -1
debug1: identity file /root/.ssh/id_xmss type -1
debug1: identity file /root/.ssh/id_xmss-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_8.6
debug1: Remote protocol version 2.0, remote software version Cisco-1.25
debug1: compat_banner: match: Cisco-1.25 pat Cisco-1.* compat 0x60000000
debug2: fd 3 setting O_NONBLOCK
debug1: Authenticating to 10.28.11.41:22 as 'root'
debug1: load_hostkeys: fopen /root/.ssh/known_hosts2: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts2: No such file or directory
debug3: order_hostkeyalgs: no algorithms matched; accept original
debug3: send packet: type 20
debug1: SSH2_MSG_KEXINIT sent
Connection closed by 10.28.11.41 port 22
```
ssh info (the other one):
```
[root@localhost system]# ssh -vvv 10.28.11.33
OpenSSH_8.6p1, OpenSSL 1.1.1k  25 Mar 2021
debug1: Reading configuration data /etc/ssh/ssh_config
debug2: resolve_canonicalize: hostname 10.28.11.33 is address
debug3: expanded UserKnownHostsFile '~/.ssh/known_hosts' -> '/root/.ssh/known_hosts'
debug3: expanded UserKnownHostsFile '~/.ssh/known_hosts2' -> '/root/.ssh/known_hosts2'
debug1: Authenticator provider $SSH_SK_PROVIDER did not resolve; disabling
debug3: ssh_connect_direct: entering
debug1: Connecting to 10.28.11.33 [10.28.11.33] port 22.
debug3: set_sock_tos: set socket 3 IP_TOS 0x48
debug1: Connection established.
debug1: identity file /root/.ssh/id_rsa type -1
debug1: identity file /root/.ssh/id_rsa-cert type -1
debug1: identity file /root/.ssh/id_dsa type -1
debug1: identity file /root/.ssh/id_dsa-cert type -1
debug1: identity file /root/.ssh/id_ecdsa type -1
debug1: identity file /root/.ssh/id_ecdsa-cert type -1
debug1: identity file /root/.ssh/id_ecdsa_sk type -1
debug1: identity file /root/.ssh/id_ecdsa_sk-cert type -1
debug1: identity file /root/.ssh/id_ed25519 type -1
debug1: identity file /root/.ssh/id_ed25519-cert type -1
debug1: identity file /root/.ssh/id_ed25519_sk type -1
debug1: identity file /root/.ssh/id_ed25519_sk-cert type -1
debug1: identity file /root/.ssh/id_xmss type -1
debug1: identity file /root/.ssh/id_xmss-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_8.6
debug1: Remote protocol version 2.0, remote software version Cisco-1.25
debug1: compat_banner: match: Cisco-1.25 pat Cisco-1.* compat 0x60000000
debug2: fd 3 setting O_NONBLOCK
debug1: Authenticating to 10.28.11.33:22 as 'root'
debug1: load_hostkeys: fopen /root/.ssh/known_hosts2: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts2: No such file or directory
debug3: order_hostkeyalgs: no algorithms matched; accept original
debug3: send packet: type 20
debug1: SSH2_MSG_KEXINIT sent
Connection closed by 10.28.11.33 port 22
```

**Comment by zhanglinqiang** [comment]
It looks like the error message is the same as the other issue #25

**Comment by norrisjeremy** [comment]
Hi @zhanglinqiang,

Do you know if you are able to successfully log into these Cisco switches with the `ssh` utility itself?
The `ssh -vvv` output you provided seems to indicate the TCP connection is being closed immediately after the `KEXINIT` packet is sent by the client:
```
debug3: send packet: type 20
debug1: SSH2_MSG_KEXINIT sent
Connection closed by 10.28.11.33 port 22
```

My suspicion is that there may be some sort of firewall between the Cisco switches and your workstation or server on which you are issuing the `ssh -vvv` commands (or running your JSch application), which is prematurely closing the TCP session (by injecting a TCP FIN packet into the TCP session).

It might also be helpful if you can perform a `tcpdump` locally wherever you are running your JSch application (or running the `ssh -vvv` commands) to see the exact raw TCP packets being sent and received.

Thanks,
Jeremy

**Comment by zhanglinqiang** [comment]
> Hi @zhanglinqiang,
> 
> Do you know if you are able to successfully log into these Cisco switches with the `ssh` utility itself? The `ssh -vvv` output you provided seems to indicate the TCP connection is being closed immediately after the `KEXINIT` packet is sent by the client:
> 
> ```
> debug3: send packet: type 20
> debug1: SSH2_MSG_KEXINIT sent
> Connection closed by 10.28.11.33 port 22
> ```
> 
> My suspicion is that there may be some sort of firewall between the Cisco switches and your workstation or server on which you are issuing the `ssh -vvv` commands (or running your JSch application), which is prematurely closing the TCP session (by injecting a TCP FIN packet into the TCP session).
> 
> It might also be helpful if you can perform a `tcpdump` locally wherever you are running your JSch application (or running the `ssh -vvv` commands) to see the exact raw TCP packets being sent and received.
> 
> Thanks, Jeremy

Sorry, it was really the network environment caused the failure. Thanks again for you help.
